### PR TITLE
Implemented a "local" CloudProvider

### PR DESCRIPTION
**Pull request by overheadhunter**
Added `LocalFsCloudProvider` which can be used for testing against the CloudProvider API on downstream dependencies.

Note that the unit test doesn't cover all cases yet, such as `NoSuchFileExceptions` when reading from non-existing files.

### CODE REVIEW

File: src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	private static final Path ABS_ROOT = Path.of("/");
+
+	private final Path root;
+
+	public LocalFsCloudProvider(Path root) {
+		this.root = root;
+	}
+
+	private Path resolve(Path cloudPath) {
+		Path relPath = ABS_ROOT.relativize(cloudPath);
```
Comment (SailReal) [inline src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java:44]: Could be `var` using type inference

File: src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		var relPath = root.relativize(fullPath);
+		var type = attr.isDirectory() ? CloudItemType.FOLDER : attr.isRegularFile() ? CloudItemType.FILE : CloudItemType.UNKNOWN;
+		var modifiedDate = Optional.of(attr.lastModifiedTime().toInstant());
+		var size = Optional.of(attr.size());
+		return new CloudItemMetadata(relPath.getFileName().toString(), ABS_ROOT.resolve(relPath), type, modifiedDate, size);
+	}
+
+	@Override
+	public CompletionStage<CloudItemMetadata> itemMetadata(Path node) {
+		Path path = resolve(node);
```
Comment (SailReal) [inline src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java:58]: Could be `var` using type inference

File: src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+			var metadata = createMetadata(path, attr);
+			return CompletableFuture.completedFuture(metadata);
+		} catch (IOException e) {
+			return CompletableFuture.failedFuture(e);
+		}
+	}
+
+	@Override
+	public CompletionStage<CloudItemList> list(Path folder, Optional<String> pageToken) {
+		Path folderPath = resolve(folder);
```
Comment (SailReal) [inline src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java:70]: Could be `var` using type inference

File: src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+		try {
+			List<CloudItemMetadata> items = new ArrayList<>();
```
Comment (SailReal) [inline src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java:72]: Could be `var` using type inference

File: src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+			});
+			return CompletableFuture.completedFuture(new CloudItemList(items, Optional.empty()));
+		} catch (IOException e) {
+			return CompletableFuture.failedFuture(e);
+		}
+	}
+
+	@Override
+	public CompletionStage<InputStream> read(Path file, long offset, long count, ProgressListener progressListener) {
+		Path filePath = resolve(file);
```
Comment (SailReal) [inline src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java:89]: Could be `var` using type inference

File: src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+			ch.position(offset);
+			return CompletableFuture.completedFuture(ByteStreams.limit(Channels.newInputStream(ch), count));
+		} catch (IOException e) {
+			return CompletableFuture.failedFuture(e);
+		}
+	}
+
+	@Override
+	public CompletionStage<CloudItemMetadata> write(Path file, boolean replace, InputStream data, ProgressListener progressListener) {
+		Path filePath = resolve(file);
```
Comment (SailReal) [inline src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java:101]: Could be `var` using type inference

File: src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+			var metadata = new CloudItemMetadata(file.getFileName().toString(), file, CloudItemType.FILE, Optional.of(modifiedDate), Optional.of(size));
+			return CompletableFuture.completedFuture(metadata);
+		} catch (IOException e) {
+			return CompletableFuture.failedFuture(e);
+		}
+	}
+
+	@Override
+	public CompletionStage<Path> createFolder(Path folder) {
+		Path folderPath = resolve(folder);
```
Comment (SailReal) [inline src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java:118]: Could be `var` using type inference

File: src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+			Files.createDirectory(folderPath);
+			return CompletableFuture.completedFuture(folder);
+		} catch (IOException e) {
+			return CompletableFuture.failedFuture(e);
+		}
+	}
+
+	@Override
+	public CompletionStage<Void> delete(Path node) {
+		Path path = resolve(node);
```
Comment (SailReal) [inline src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java:129]: Could be `var` using type inference

File: src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+			return CompletableFuture.completedFuture(null);
+		} catch (IOException e) {
+			return CompletableFuture.failedFuture(e);
+		}
+	}
+
+	@Override
+	public CompletionStage<Path> move(Path source, Path target, boolean replace) {
+		Path src = resolve(source);
+		Path dst = resolve(target);
```
Comment (SailReal) [inline src/main/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProvider.java:141]: Could be `var` using type inference

File: src/test/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProviderTest.java [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+		try (var in = Assertions.assertTimeoutPreemptively(Duration.ofSeconds(1), () -> result.toCompletableFuture().get())) {
+			var allBytes = in.readAllBytes();
+			Assertions.assertArrayEquals("o w".getBytes(), allBytes);
+		}
+	}
+
+	@Test
+	@DisplayName("write to /file (non-existing)")
+	public void testWriteToNewFile() throws IOException {
+		//Files.write(root.resolve("file"), "hello world".getBytes());
```
Comment (SailReal) [inline src/test/java/org/cryptomator/cloudaccess/localfs/LocalFsCloudProviderTest.java]: Can be deleted

**Review by SailReal — COMMENTED** [review]
LGTM 👍 , just a few comments about type inference 😅

**Comment by overheadhunter** [comment]
I tend to use `var` only in places where explicit types are very long _or_ to keep a whole bunch of declarations clean where the variable name is unambiguous.

For short types, superclasses/interfaces as well as places where I need developers (others as well of my future self) to be perfectly aware of the type, I prefer explicit type declarations.

So in other words: I don't have any strict rule when to use `var` yet.

**Review by SailReal — APPROVED** [review]

**Pull request merged**